### PR TITLE
[pallet-transaction-storage] Improved `check_proof` check + tests + docs

### PR DESCRIPTION
**Pull request by bkontur**
**This PR:**

* Fixes `check_proof` and its `binary_search_by_key` chunk
* Adds the `ensure_chunk_proof_works` test, which covers all possible chunk index build/verify proof roundtrips (to catch all corner cases)
* Improves docs around `pallet-transaction-storage`

### CODE REVIEW

File: substrate/primitives/transaction-storage-proof/src/lib.rs [hunk 1]
```diff
@@ -182,19 +194,20 @@ pub mod registration {
...
 		transactions: Vec<Vec<u8>>,
 	) -> Result<TransactionStorageProof, Error> {
-		let mut db = sp_trie::MemoryDB::<Hasher>::default();
+		// Get total chunks, we will need it to generate a random chunk index.
+		let total_chunks: u32 = transactions.iter().map(|t| num_chunks(t.len() as u32)).sum();
+		if total_chunks.is_zero() {
+			const MSG: &str = "No chunks to build proof for.";
+			return Err(Error::Application(Box::from(MSG)));
+		}
+		let target_chunk_index = random_chunk(random_hash.as_ref(), total_chunks);
```
Comment (bkontur) [inline substrate/primitives/transaction-storage-proof/src/lib.rs]: clippy
```suggestion
		let target_chunk_index = random_chunk(random_hash, total_chunks);
```

File: prdoc/pr_10153.prdoc [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+    **This PR:**
+
+    * Fixes `check_proof` and its `binary_search_by_key` chunk
+    * Adds the `ensure_chunk_proof_works` test, which covers all possible chunk index build/verify proof roundtrips (to catch all corner cases)
+    * Improves docs around `pallet-transaction-storage`
+crates:
+- name: pallet-transaction-storage
+  bump: patch
+- name: sp-transaction-storage-proof
+  bump: patch
```
Comment (bkontur) [inline prdoc/pr_10153.prdoc]:
```suggestion
  bump: minor
```

File: substrate/frame/transaction-storage/src/lib.rs [hunk 3]
```diff
@@ -466,5 +450,60 @@ pub mod pallet {
 			T::FeeDestination::on_unbalanced(credit);
 			Ok(())
 		}
+
+		pub(crate) fn ensure_chunk_proof(
```
Comment (bkchr) [inline substrate/frame/transaction-storage/src/lib.rs]:
```suggestion
		pub(crate) fn verify_chunk_proof(
```

Comment (bkchr) [reply substrate/frame/transaction-storage/src/lib.rs]: And at least one line of docs :)

File: substrate/frame/transaction-storage/src/lib.rs [hunk 4]
```diff
@@ -301,39 +300,24 @@ pub mod pallet {
 		) -> DispatchResultWithPostInfo {
 			ensure_none(origin)?;
 			ensure!(!ProofChecked::<T>::get(), Error::<T>::DoubleCheck);
+
+			// Get the target block metadata.
 			let number = frame_system::Pallet::<T>::block_number();
 			let period = StoragePeriod::<T>::get();
 			let target_number = number.saturating_sub(period);
 			ensure!(!target_number.is_zero(), Error::<T>::UnexpectedProof);
 			let total_chunks = ChunkCount::<T>::get(target_number);
```
Comment (bkchr) [inline substrate/frame/transaction-storage/src/lib.rs]: `ChunkCount` is not really required, because this is just `transactions.last().map_or(0, |t| t.block_chunks)`

Comment (bkchr) [reply substrate/frame/transaction-storage/src/lib.rs]: So this could be removed.

File: substrate/primitives/transaction-storage-proof/src/lib.rs [hunk 5]
```diff
@@ -114,14 +116,25 @@ impl sp_inherents::InherentDataProvider for InherentDataProvider {
 }
 
 /// A utility function to extract a chunk index from the source of randomness.
+///
+/// Note: The caller must ensure that `total_chunks` is not 0.
```
Comment (bkchr) [inline substrate/primitives/transaction-storage-proof/src/lib.rs]:
```suggestion
/// # Panics 
///
/// This function panics if `total_chunks` is `0`.
```

File: substrate/primitives/transaction-storage-proof/src/lib.rs [hunk 6]
```diff
@@ -17,6 +17,8 @@
 
 //! Storage proof primitives. Contains types and basic code to extract storage
 //! proofs for indexed transactions.
+//!
+//! Note: We use `u32` for both `total_chunks` and the chunk index.
```
Comment (bkchr) [inline substrate/primitives/transaction-storage-proof/src/lib.rs]:
```suggestion
```

Don't really get what this should tell us.

File: substrate/primitives/transaction-storage-proof/src/lib.rs [hunk 7]
```diff
@@ -182,19 +194,20 @@ pub mod registration {
 		random_hash: &[u8],
 		transactions: Vec<Vec<u8>>,
 	) -> Result<TransactionStorageProof, Error> {
-		let mut db = sp_trie::MemoryDB::<Hasher>::default();
+		// Get total chunks, we will need it to generate a random chunk index.
+		let total_chunks: u32 = transactions.iter().map(|t| num_chunks(t.len() as u32)).sum();
+		if total_chunks.is_zero() {
+			const MSG: &str = "No chunks to build proof for.";
+			return Err(Error::Application(Box::from(MSG)));
+		}
```
Comment (bkchr) [inline substrate/primitives/transaction-storage-proof/src/lib.rs:205]: Why is this now an error? This will make the block production fail. Before it was just not providing any proof, which is the sensible thing to do here.

File: substrate/frame/transaction-storage/src/lib.rs (hunk 4 above)
Comment (bkontur) [reply substrate/frame/transaction-storage/src/lib.rs]: > `ChunkCount` is not really required, because this is just `transactions.last().map_or(0, |t| t.block_chunks)`

Yes, exactly, I also realized that `ChunkCount` is not really needed and redundant, but I thought that it is kind of double-checking. 

@bkchr so you are suggesting to remove `ChunkCount` storage item?

File: substrate/primitives/transaction-storage-proof/src/lib.rs (hunk 6 above)
Comment (bkontur) [reply substrate/primitives/transaction-storage-proof/src/lib.rs]: > Don't really get what this should tell us.

I wanted to reuse `num_chunks` / `random_chunk` but in both algos (build_proof / verify_chunk_proof) we are using couple of casting usize->u64, usize->u32, but at the end, the only thing that matters is `total_chunks: u32` and `fn encode_index(input: u32)`, but ok, let's remove

File: substrate/frame/transaction-storage/src/lib.rs (hunk 3 above)
Comment (bkontur) [reply substrate/frame/transaction-storage/src/lib.rs]: Fixed by: https://github.com/paritytech/polkadot-sdk/pull/10153/commits/513fa70a1b8d29ed9c582205fe1a2d7e4363011d

File: substrate/frame/transaction-storage/src/lib.rs (hunk 4 above)
Comment (bkchr) [reply substrate/frame/transaction-storage/src/lib.rs]: Yes

File: substrate/primitives/transaction-storage-proof/src/lib.rs (hunk 6 above)
Comment (bkchr) [reply substrate/primitives/transaction-storage-proof/src/lib.rs]: Wouldn't it make more sense to do:
```
pub type ChunkIndex = u32;
```

I see purpose in this, but not this comment :)

File: substrate/primitives/transaction-storage-proof/src/lib.rs (hunk 7 above)
Comment (bkontur) [reply substrate/primitives/transaction-storage-proof/src/lib.rs:205]: > Why is this now an error? This will make the block production fail. Before it was just not providing any proof, which is the sensible thing to do here.

oh, good point, thank you, you're right, my bad, please check here: https://github.com/paritytech/polkadot-sdk/pull/10153/commits/04cb7d7245773f4bd90585908a5c328ab00e973b

I also tried to remove that `.unwrap(` and change return type to `Result<Option`, but I am not sure about the end (looks ok-ish and should not happen):
```
debug_assert!(false, "No chunk matched the target_chunk_index; logic error");
Ok(None)
```
@bkchr What would you suggest here?

---
I also talked to @arkpar and added an early return when the chunk is found, since we don’t need to process the subsequent transactions.

File: substrate/primitives/transaction-storage-proof/src/lib.rs (hunk 6 above)
Comment (bkontur) [reply substrate/primitives/transaction-storage-proof/src/lib.rs]: > pub type ChunkIndex = u32;

Yeah, I was also thinking about this, but didn't want to complicate :D
I will change that :)

Comment (bkontur) [reply substrate/primitives/transaction-storage-proof/src/lib.rs]: `ChunkIndex` here: https://github.com/paritytech/polkadot-sdk/pull/10153/commits/23eee5b13c9610c13dd650744afd58e0453a4f61

File: substrate/frame/transaction-storage/src/lib.rs (hunk 4 above)
Comment (bkontur) [reply substrate/frame/transaction-storage/src/lib.rs]: > Yes

`ChunkCount` is removed here: https://github.com/paritytech/polkadot-sdk/pull/10153/commits/7d1b78a02f903a0cef6b85100af57d32392607c2

> ...I thought that it is kind of double-checking.

I might be wrong, but it seems `ChunkCount` wasn’t for double-checking — it was an optimization for `on_finalize` to avoid decoding the entire `BoundedVec<TransactionInfo` just to read the last `TransactionInfo.block_chunks`. Here: https://github.com/paritytech/polkadot-sdk/pull/10153/commits/7d1b78a02f903a0cef6b85100af57d32392607c2#diff-c2ba025b8a704674283b2fb52ac5f67fa341ff24832c2a9b24afa792a4bd2c61L182-R198.

@bkchr wdyt? Revert-and-keep `ChunkCount` or `Transactions::<T>::decode_len` fix is enough or other solution?

File: prdoc/pr_10153.prdoc [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+- audience: Runtime Dev
+  description: |-
+    **This PR:**
+
+    * Fixes `check_proof` and its `binary_search_by_key` chunk
+    * Adds the `verify_chunk_proof_works` test, which covers all possible chunk index build/verify proof roundtrips (to catch all corner cases)
+    * Improves docs around `pallet-transaction-storage`
+crates:
+- name: pallet-transaction-storage
+  bump: patch
```
Comment (bkontur) [inline prdoc/pr_10153.prdoc]:
```suggestion
  bump: major
```

File: substrate/frame/transaction-storage/src/lib.rs [hunk 9]
```diff
@@ -80,12 +80,20 @@ pub struct TransactionInfo {
...
+	/// total chunks value.
+	block_chunks: ChunkIndex,
 }
 
-fn num_chunks(bytes: u32) -> u32 {
-	(bytes as u64).div_ceil(CHUNK_SIZE as u64) as u32
+impl TransactionInfo {
+	/// Get the number of total chunks.
+	///
+	/// See [`TransactionInfo::block_chunks`].
```
Comment (bkontur) [inline substrate/frame/transaction-storage/src/lib.rs]:
```suggestion
	/// See `TransactionInfo::block_chunks`.
```

File: substrate/frame/transaction-storage/src/lib.rs [hunk 10]
```diff
@@ -180,15 +194,24 @@ pub mod pallet {
...
+
+					target_number.is_zero() || {
+						// An empty block means no transactions were stored, relying on the fact
+						// below that we store transactions only if they contain chunks.
+						//
+						// Optimization: do not decode transactions; just get their length.
+						// Otherwise, it would look like:
+						// `Transactions::<T>::get(target_number).map(|txs|
+						// TransactionInfo::total_chunks(&txs)).unwrap_or_default()`
+						Transactions::<T>::decode_len(target_number).unwrap_or_default() == 0
```
Comment (bkchr) [inline substrate/frame/transaction-storage/src/lib.rs]:
```suggestion
						!Transactions::<T>::exists(target_number)
```

This should be enough? We may need to check that we don't write if there isn't anything.

Comment (bkontur) [reply substrate/frame/transaction-storage/src/lib.rs]: > This should be enough?

Oh, nice, yes, looks like this `exists` should be enough, perfect.

> We may need to check that we don't write if there isn't anything.

Yes, exactly, we write only when we have something, just few lines below:

```
			if total_chunks != 0 {
				Transactions::<T>::insert(n, transactions);
			}
```

File: substrate/frame/transaction-storage/src/lib.rs [hunk 11]
```diff
@@ -180,15 +194,24 @@ pub mod pallet {
...
+
+					target_number.is_zero() || {
+						// An empty block means no transactions were stored, relying on the fact
+						// below that we store transactions only if they contain chunks.
+						//
+						// Optimization: do not decode transactions; just get their length.
+						// Otherwise, it would look like:
+						// `Transactions::<T>::get(target_number).map(|txs|
+						// TransactionInfo::total_chunks(&txs)).unwrap_or_default()`
+						!Transactions::<T>::exists(target_number)
```
Comment (bkontur) [inline substrate/frame/transaction-storage/src/lib.rs]:
```suggestion
						!Transactions::<T>::contains_key(target_number)
```

File: substrate/frame/transaction-storage/src/lib.rs [hunk 12]
```diff
@@ -180,15 +194,24 @@ pub mod pallet {
...
-					target_number.is_zero() || ChunkCount::<T>::get(target_number) == 0
+
+					target_number.is_zero() || {
+						// An empty block means no transactions were stored, relying on the fact
+						// below that we store transactions only if they contain chunks.
+						//
+						// Optimization: do not decode transactions; just get their length.
+						// Otherwise, it would look like:
+						// `Transactions::<T>::get(target_number).map(|txs|
+						// TransactionInfo::total_chunks(&txs)).unwrap_or_default()`
```
Comment (bkontur) [inline substrate/frame/transaction-storage/src/lib.rs]:
```suggestion
```

File: substrate/frame/transaction-storage/src/lib.rs (hunk 4 above)
Comment (bkontur) [reply substrate/frame/transaction-storage/src/lib.rs]: Fixed by: https://github.com/paritytech/polkadot-sdk/pull/10153#discussion_r2482822736
(`Transactions::<T>::contains_key`)

File: substrate/frame/transaction-storage/src/lib.rs [hunk 13]
```diff
@@ -80,12 +80,20 @@ pub struct TransactionInfo {
...
+	/// total chunks value.
+	block_chunks: ChunkIndex,
 }
 
-fn num_chunks(bytes: u32) -> u32 {
-	(bytes as u64).div_ceil(CHUNK_SIZE as u64) as u32
+impl TransactionInfo {
+	/// Get the number of total chunks.
+	///
+	/// See `TransactionInfo::block_chunks`.
```
Comment (bkchr) [inline substrate/frame/transaction-storage/src/lib.rs]:
```suggestion
	/// See [`TransactionInfo::block_chunks`].
```

File: substrate/frame/transaction-storage/src/lib.rs [hunk 14]
```diff
@@ -161,16 +169,22 @@ pub mod pallet {
 	#[pallet::hooks]
 	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
 		fn on_initialize(n: BlockNumberFor<T>) -> Weight {
+			let mut weight = Weight::zero();
```
Comment (bkchr) [inline substrate/frame/transaction-storage/src/lib.rs:173]: Maybe good idea to create an issue for this, to have proper benchmarking.

File: substrate/primitives/transaction-storage-proof/src/lib.rs [hunk 15]
```diff
@@ -207,33 +219,39 @@ pub mod registration {
...
+				// so there is no need to waste time calculating the subsequent transactions.
+				return Ok(Some(TransactionStorageProof {
+					proof: chunk_proof,
+					chunk: target_chunk,
+				}));
 			}
 		}
 
-		Ok(TransactionStorageProof { proof: chunk_proof, chunk: target_chunk.unwrap() })
+		debug_assert!(false, "No chunk matched the target_chunk_index; logic error");
```
Comment (bkchr) [inline substrate/primitives/transaction-storage-proof/src/lib.rs]: Probably returning an error would be better as I assume that the block would fail to produce any way?

File: substrate/frame/transaction-storage/src/lib.rs (hunk 13 above)
Comment (bkontur) [reply substrate/frame/transaction-storage/src/lib.rs]: I had it like this initially, but rust-doc job complained: paritytech/polkadot-sdk/actions/runs/18924478339/job/54028604239#step:4:2929 and I didn't want to add `pub` or `#![allow(rustdoc::private_intra_doc_links)]`.

Rewritten here: https://github.com/paritytech/polkadot-sdk/pull/10153/commits/fca0b92d5a4dcf3a8f0a79f756018a8da5284d07

File: substrate/primitives/transaction-storage-proof/src/lib.rs (hunk 15 above)
Comment (bkontur) [reply substrate/primitives/transaction-storage-proof/src/lib.rs]: > Probably returning an error would be better as I assume that the block would fail to produce any way?

Yes, right.
(This is part node part, so we could potentially use `unreachable!()` panic, but let’s start gracefully with just Err)

Fixed by: https://github.com/paritytech/polkadot-sdk/pull/10153/commits/97cf2fa400ad68bec246254b8b0f1201096cc723

File: substrate/frame/transaction-storage/src/lib.rs (hunk 14 above)
Comment (bkontur) [inline substrate/frame/transaction-storage/src/lib.rs:173]:
```suggestion
			// TODO: https://github.com/paritytech/polkadot-sdk/issues/10203 - Replace this with benchmarked weights.
			let mut weight = Weight::zero();
```

Comment (bkontur) [reply substrate/frame/transaction-storage/src/lib.rs:173]: Good point, issue created: https://github.com/paritytech/polkadot-sdk/issues/10203
